src/Servers/Kestrel/Core/src/Internal/Http/Http1ContentLengthMessageBody.cs

```diff
@@ -0,0 +1,295 @@
+            : base(context)
+        {
+            RequestKeepAlive = keepAlive;
+            _contentLength = contentLength;
+            _inputLength = _contentLength;
+        }
+
+        public override async ValueTask<ReadResult> ReadAsync(CancellationToken cancellationToken = default)
+        {
+            ThrowIfCompleted();
+
+            if (_inputLength == 0)
+            {
+                _readResult = new ReadResult(default, isCanceled: false, isCompleted: true);
+                return _readResult;
+            }
+
+            TryStart();
+
+            // The while(true) loop is required because the Http1 connection calls CancelPendingRead to unblock
+            // the call to StartTimingReadAsync to check if the request timed out.
+            // However, if the user called CancelPendingRead, we want that to return a canceled ReadResult
+            // We internally track an int for that.
+            while (true)
+            {
+                // This isn't great. The issue is that TryRead can get a canceled read result
+                // which is unknown to StartTimingReadAsync. 
+                if (_context.RequestTimedOut)
+                {
+                    BadHttpRequestException.Throw(RequestRejectionReason.RequestBodyTimeout);
+                }
+
+                _readResult = await StartTimingReadAsync(cancellationToken);
+
+                if (_context.RequestTimedOut)
+                {
+                    Debug.Assert(_readResult.IsCanceled);
+                    BadHttpRequestException.Throw(RequestRejectionReason.RequestBodyTimeout);
+                }
+
+                if (_readResult.IsCanceled)
+                {
+                    if (Interlocked.CompareExchange(ref _userCanceled, 0, 1) == 1)
+                    {
+                        // Ignore the readResult if it wasn't by the user.
+                        break;
+                    }
+                }
+
+                var readableBuffer = _readResult.Buffer;
+                var readableBufferLength = readableBuffer.Length;
+                StopTimingRead(readableBufferLength);
+
+                if (_readResult.IsCompleted)
+                {
+                    // OnInputOrOutputCompleted() is an idempotent method that closes the connection. Sometimes
+                    // input completion is observed here before the Input.OnWriterCompleted() callback is fired,
+                    // so we call OnInputOrOutputCompleted() now to prevent a race in our tests where a 400
+                    // response is written after observing the unexpected end of request content instead of just
+                    // closing the connection without a response as expected.
+                    _context.OnInputOrOutputCompleted();
+
+                    BadHttpRequestException.Throw(RequestRejectionReason.UnexpectedEndOfRequestContent);
+                }
```

Comment:
It would be nice if this along with the other shared ReadResult validation were moved into shared methods instead of duplicated.